### PR TITLE
Feature/compute_conditions

### PR DESCRIPTION
**Pull request by aabdulwahed**
## Description
**PLEASE DO NOTE THAT THIS IS NOT THE FINAL TEMPLATE AND IT WILL BE `accumulated` OVER TIME**
- Initial implementation of [on-premise compute](https://github.com/oceanprotocol/dev-ocean/blob/master/doc/architecture/computing-services.md#flow) conditions.
 
- **Future updates:** 
  - add new condition indicates the did registration of a new derived asset by publisher which in turn will assign the ownership of the derived asset to the data-scientist.
  - verification game conditions (in terms of dispute condition). This might need off-chain verifier implementation.
 
## Is this PR related with an open issue?

Related to Issue [#186](https://github.com/oceanprotocol/ocean/issues/186) and [#134](https://github.com/oceanprotocol/ocean/issues/134)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [x] Follows the code style of this project.
- [x] Tests Cover Changes
- [x] Documentation

### CODE REVIEW

File: contracts/SLA/ComputeConditions.sol [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import './ServiceAgreement.sol';
+
+/// @title On-premise compute conditions
+/// @author Ocean Protocol Team
+/// @notice This contract is WIP, don't use if for production
+/// @dev All function calls are currently implement without side effects
+
+contract ComputeConditions {
+
+    struct UploadProof {
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: UploadProof sounds like an action. It should be called ProofOfUpload

File: contracts/SLA/ComputeConditions.sol [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        bool created;
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: either isCreated or exists so you can do statements like (ProofOfUpload.exists) thats way easier to understand.

File: contracts/SLA/ComputeConditions.sol [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        bool valid;
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: isValid

File: contracts/SLA/ComputeConditions.sol [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        bool locked;
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: isLocked

File: contracts/SLA/ComputeConditions.sol [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        address dataScientist;
+        bytes32 algorithmHash;
+        bytes signature;
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: algorithmHashSignature

File: contracts/SLA/ComputeConditions.sol [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+/// @notice This contract is WIP, don't use if for production
+/// @dev All function calls are currently implement without side effects
+
+contract ComputeConditions {
+
+    struct UploadProof {
+        bool created;
+        bool valid;
+        bool locked;
+        address dataScientist;
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: dataConsumer

File: contracts/SLA/ComputeConditions.sol [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    ServiceAgreement private serviceAgreementStorage;
+    mapping (bytes32 => UploadProof) proofs;
+
+    //events
+    event SignatureSubmitted(bytes32 serviceAgreementId, address dataScientist, address publisher, bool state);
+    event HashSubmitted(bytes32 serviceAgreementId, address dataScientist, address publisher, bool state);
+    event ValidUploadProof(bytes32 serviceAgreementId, address dataScientist, address publisher, bool state);
+
+    modifier onlyDataScientist(bytes32 serviceAgreementId) {
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: onlyDataConsumer

File: contracts/SLA/ComputeConditions.sol [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        address dataScientist;
+        bytes32 algorithmHash;
+        bytes signature;
+    }
+
+    ServiceAgreement private serviceAgreementStorage;
+    mapping (bytes32 => UploadProof) proofs;
+
+    //events
+    event SignatureSubmitted(bytes32 serviceAgreementId, address dataScientist, address publisher, bool state);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: HashSignatureSubmitted

File: contracts/SLA/ComputeConditions.sol [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    /// @notice submitSignature is called only by the data-scientist address.
+    /// @dev At first It checks if the proof state is created or not then checks that the hash
+    /// has been submitted by the publisher in order to call fulfillUpload. This preserves
+    /// the message integrity and proof that both parties agree on the same algorithm file/s
+    /// @param serviceAgreementId , the service level agreement Id
+    /// @param signature data scientist signature = signed_hash(uploaded_algorithm_file/s)
+    function submitSignature(bytes32 serviceAgreementId, bytes signature) public onlyDataScientist(serviceAgreementId) returns(bool) {
+        if(proofs[serviceAgreementId].created){
+            if(proofs[serviceAgreementId].locked) { // avoid race conditions
+                emit SignatureSubmitted(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), false);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: This is hard to implement for a consumer like squid. You cannot tell by the event wether this method has succeeded or not. I recommend emitting two events.

File: contracts/SLA/ComputeConditions.sol [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    /// @param state get be used fo input value hash for this condition indicating the state of verification
+    function fulfillUpload(bytes32 serviceAgreementId, bool state) public onlyStakeholders(serviceAgreementId) returns(bool) {
+        bytes32 condition = serviceAgreementStorage.getConditionByFingerprint(serviceAgreementId, address(this), this.fulfillUpload.selector);
+        if (serviceAgreementStorage.hasUnfulfilledDependencies(serviceAgreementId, condition)){
+            emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), false);
+            return false;
+        }
+
+        if (serviceAgreementStorage.getConditionStatus(serviceAgreementId, condition) == 1) {
+            emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), true);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: ProofOfUploadValid

File: contracts/SLA/ComputeConditions.sol [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    /// @notice fulfillUpload is called by anyone of the stakeholders [publisher or data scientist]
+    /// @dev check if there are unfulfilled dependency condition, if false, it verifies the signature
+    /// using the submitted hash (by publisher), the signature (by data scientist) then call
+    /// fulfillCondition in service level agreement storage contract
+    /// @param serviceAgreementId the service level agreement Id
+    /// @param state get be used fo input value hash for this condition indicating the state of verification
+    function fulfillUpload(bytes32 serviceAgreementId, bool state) public onlyStakeholders(serviceAgreementId) returns(bool) {
+        bytes32 condition = serviceAgreementStorage.getConditionByFingerprint(serviceAgreementId, address(this), this.fulfillUpload.selector);
+        if (serviceAgreementStorage.hasUnfulfilledDependencies(serviceAgreementId, condition)){
+            emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), false);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: ProofOfUploadInvalid

File: contracts/SLA/ComputeConditions.sol [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        if (serviceAgreementStorage.getConditionStatus(serviceAgreementId, condition) == 1) {
+            emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), true);
+            return true;
+        }
+
+        if(proofs[serviceAgreementId].dataScientist == ECDSA.recover(ECDSA.toEthSignedMessageHash(proofs[serviceAgreementId].algorithmHash), proofs[serviceAgreementId].signature)) {
+            serviceAgreementStorage.fulfillCondition(serviceAgreementId, this.fulfillUpload.selector, keccak256(abi.encodePacked(state)));
+            emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), true);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: ProofOfUploadValid

File: contracts/SLA/ComputeConditions.sol [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+            proofs[serviceAgreementId].valid = true;
+            return true;
+        }
+        emit ValidUploadProof(serviceAgreementId, serviceAgreementStorage.getServiceAgreementConsumer(serviceAgreementId), serviceAgreementStorage.getAgreementPublisher(serviceAgreementId), false);
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: ProofOfUploadInvalid

File: contracts/SLA/ComputeConditions.sol [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        serviceAgreementStorage = ServiceAgreement(serviceAgreementAddress);
+    }
+
+    /// @notice submitSignature is called only by the data-scientist address.
+    /// @dev At first It checks if the proof state is created or not then checks that the hash
+    /// has been submitted by the publisher in order to call fulfillUpload. This preserves
+    /// the message integrity and proof that both parties agree on the same algorithm file/s
+    /// @param serviceAgreementId , the service level agreement Id
+    /// @param signature data scientist signature = signed_hash(uploaded_algorithm_file/s)
+    function submitSignature(bytes32 serviceAgreementId, bytes signature) public onlyDataScientist(serviceAgreementId) returns(bool) {
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: submitHashSignature

File: contracts/SLA/ComputeConditions.sol (hunk 9 above)
Comment (aabdulwahed) [reply contracts/SLA/ComputeConditions.sol]: The consumer receives the transaction receipt where the function returns T/F. You don't have to rely on the event.

**Comment by H34D** [comment]
looks good, thanks @aabdulwahed

**Pull request merged**